### PR TITLE
[Webhook] Add Mailchimp webhook

### LINKED ISSUE #50285: [Webhook] Add webhook implementation for Mailchimp
### Description

Since the [Webhook component](https://github.com/symfony/symfony/pull/48542) is available would be nice if we add implementation for Mailchimp some information: https://mailchimp.com/developer/transactional/docs/webhooks/

I can work on this as i did implemented already in our Symfony project.

### Example

_No response_

### PR DESCRIPTION
**Pull request by JohJohan**
| Q             | A
| ------------- | ---
| Branch?       | 7.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #50285 
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/20286

This will add Mailchimp webhook see https://mailchimp.com/developer/transactional/docs/webhooks/

### CODE REVIEW

File: src/Symfony/Component/Mailer/Bridge/Mailchimp/Tests/Webhook/MailchimpRequestParserTest.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+{
+    protected function createRequestParser(): RequestParserInterface
+    {
+        return new MailchimpRequestParser(new MailchimpPayloadConverter());
+    }
+
+    protected function createRequest(string $payload): Request
+    {
+        $request = parent::createRequest($payload);
+        $request->headers->set('X-Mandrill-Signature', 'fsGYPUQPD9m7+q02GVykHW80hxc=');
```
Comment (JohJohan) [inline src/Symfony/Component/Mailer/Bridge/Mailchimp/Tests/Webhook/MailchimpRequestParserTest.php]: To correctly validate the signature i have to set it based on the fixture data so it will be unique value for each fixrure. 
I had 2 implementation in mind:
1. Add a public function in the x.php file to get the signature
2. Add a extra file called something like x.signature and read the contents?
Do you have other suggestions?

**Comment by OskarStark** [comment]
Any plans to finish this PR @JohJohan ?

**Comment by JohJohan** [comment]
> Any plans to finish this PR @JohJohan ?

Yes will wrap up today. For https://github.com/symfony/symfony/pull/50324#discussion_r1193644644 i will add the values in the test.json files and unset the values.

File: src/Symfony/Component/Mailer/Bridge/Mailchimp/Tests/Webhook/MailchimpRequestParserTest.php (hunk 1 above)
Comment (OskarStark) [reply src/Symfony/Component/Mailer/Bridge/Mailchimp/Tests/Webhook/MailchimpRequestParserTest.php]: 1. sounds good to me

**Comment by ghost** [comment]
I see now MailChimp will send multiple events at once see https://mailchimp.com/developer/transactional/guides/track-respond-activity-webhooks/#handling-the-webhook-response-in-your-app

I'm thinking to make a class `MailerDeliveriesEvent`  that contains `Symfony\Component\RemoteEvent\Event\Mailer\MailerDeliveryEvent` and will add `email` to `MailerDeliveryEvent` so you can handle each event. The concume should then look like:
```php

    public function consume(RemoteEvent|MailerDeliveriesEvent $event): void
    {

        foreach ($event as $deliveryEvent) {
           // Handle the event based on email
           $deliveryEvent->getEmail()
        }
    }
```

Maybe you also have other suggestion? I dont like it that `consume` function will have the naming for 1 or multiple events and that is probably not a good idea.

File: src/Symfony/Component/Mailer/Bridge/Mailchimp/Tests/Webhook/MailchimpRequestParserTest.php (hunk 1 above)
Comment (ghost) [reply src/Symfony/Component/Mailer/Bridge/Mailchimp/Tests/Webhook/MailchimpRequestParserTest.php]: See https://github.com/symfony/symfony/pull/50324#issuecomment-1876891274

File: src/Symfony/Component/Mailer/Bridge/Mailchimp/RemoteEvent/MailchimpPayloadConverter.php [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+use Symfony\Component\RemoteEvent\Exception\ParseException;
+use Symfony\Component\RemoteEvent\PayloadConverterInterface;
+
+final class MailchimpPayloadConverter implements PayloadConverterInterface
+{
+    public function convert(array $payload): AbstractMailerEvent
+    {
+        if (!\in_array($payload['event'], ['send', 'deferral', 'hard_bounce', 'soft_bounce', 'delivered', 'open', 'click', 'spam', 'unsub', 'reject'], true)) {
+            throw new ParseException(sprintf('Unsupported event "%s".', $payload['event']));
+        }
```
Comment (alamirault) [inline src/Symfony/Component/Mailer/Bridge/Mailchimp/RemoteEvent/MailchimpPayloadConverter.php]: We can avoid this redundant condition

See MailgunPayloadConverter 
https://github.com/symfony/symfony/blob/3f2ed0fc4c41744991113185a619baf5d2f7590f/src/Symfony/Component/Mailer/Bridge/Mailgun/RemoteEvent/MailgunPayloadConverter.php#L49

File: src/Symfony/Component/Mailer/Bridge/Mailchimp/RemoteEvent/MailchimpPayloadConverter.php [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        if (\in_array($payload['event'], ['send', 'deferral', 'soft_bounce', 'hard_bounce', 'delivered', 'reject'], true)) {
+            $name = match ($payload['event']) {
+                'send' => MailerDeliveryEvent::RECEIVED,
+                'deferral', 'soft_bounce' => MailerDeliveryEvent::DEFERRED,
+                'hard_bounce' => MailerDeliveryEvent::BOUNCE,
```
Comment (alamirault) [inline src/Symfony/Component/Mailer/Bridge/Mailchimp/RemoteEvent/MailchimpPayloadConverter.php]: In BrevoPayloadConverter, soft_bounce and hard_bounce match `MailerDeliveryEvent::BOUNCE`, should we do same here ?

File: src/Symfony/Component/Mailer/Bridge/Mailchimp/RemoteEvent/MailchimpPayloadConverter.php [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    {
+        if (null !== $payload['msg']['diag']) {
+            return $payload['msg']['diag'];
+        }
+        if (null !== $payload['msg']['bounce_description']) {
+            return $payload['msg']['bounce_description'];
+        }
+        if (null !== $payload['msg']['smtp_events'] && null !== $payload['msg']['smtp_events']['diag']) {
+            return $payload['msg']['smtp_events']['diag'];
+        }
```
Comment (alamirault) [inline src/Symfony/Component/Mailer/Bridge/Mailchimp/RemoteEvent/MailchimpPayloadConverter.php]: According to https://mailchimp.com/developer/transactional/docs/webhooks/#detailed-webhook-format-responses, is an array of object right ?

**Review by alamirault — CHANGES_REQUESTED** [review]

**Comment by alamirault** [comment]
> I see now MailChimp will send multiple events at once see https://mailchimp.com/developer/transactional/guides/track-respond-activity-webhooks/#handling-the-webhook-response-in-your-app
> 
> I'm thinking to make a class `MailerDeliveriesEvent` that contains `Symfony\Component\RemoteEvent\Event\Mailer\MailerDeliveryEvent` and will add `email` to `MailerDeliveryEvent` so you can handle each event. The concume should then look like:
> 
> ```
>     public function consume(RemoteEvent|MailerDeliveriesEvent $event): void
>     {
> 
>         foreach ($event as $deliveryEvent) {
>            // Handle the event based on email
>            $deliveryEvent->getEmail()
>         }
>     }
> ```
> 
> Maybe you also have other suggestion? I dont like it that `consume` function will have the naming for 1 or multiple events and that is probably not a good idea.

Indeed, bulk event in payload is complicated with this signature
https://github.com/symfony/symfony/blob/3f2ed0fc4c41744991113185a619baf5d2f7590f/src/Symfony/Component/Webhook/Client/RequestParserInterface.php#L24-L31

(Fore this case, SendgridRequestParser consume only the first event and ignore others :fearful:)

I'm not sure if it's the best way, but changing signature (without BC break) to return an array could be cleaner
```
    /**
     * Parses an HTTP Request and converts it into a RemoteEvent.
     *
     * @return array<RemoteEvent> Returns empty if the webhook must be ignored
     *
     * @throws RejectWebhookException When the payload is rejected (signature issue, parse issue, ...)
     */
    public function parse(Request $request, #[\SensitiveParameter] string $secret): array;
```

File: src/Symfony/Component/Mailer/Bridge/Mailchimp/RemoteEvent/MailchimpPayloadConverter.php (hunk 4 above)
Comment (JohJohan) [reply src/Symfony/Component/Mailer/Bridge/Mailchimp/RemoteEvent/MailchimpPayloadConverter.php]: Your right `an array of JSON objects, each of which is an SMTP response received for the message. Each item in the array will contain the following keys: `

I will have to update to read every `diag` and concat them together

**Comment by JohJohan** [comment]
> > I see now MailChimp will send multiple events at once see https://mailchimp.com/developer/transactional/guides/track-respond-activity-webhooks/#handling-the-webhook-response-in-your-app
> > I'm thinking to make a class `MailerDeliveriesEvent` that contains `Symfony\Component\RemoteEvent\Event\Mailer\MailerDeliveryEvent` and will add `email` to `MailerDeliveryEvent` so you can handle each event. The concume should then look like:
> > ```
> >     public function consume(RemoteEvent|MailerDeliveriesEvent $event): void
> >     {
> > 
> >         foreach ($event as $deliveryEvent) {
> >            // Handle the event based on email
> >            $deliveryEvent->getEmail()
> >         }
> >     }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Maybe you also have other suggestion? I dont like it that `consume` function will have the naming for 1 or multiple events and that is probably not a good idea.
> 
> Indeed, bulk event in payload is complicated with this signature
> 
> https://github.com/symfony/symfony/blob/3f2ed0fc4c41744991113185a619baf5d2f7590f/src/Symfony/Component/Webhook/Client/RequestParserInterface.php#L24-L31
> 
> (Fore this case, SendgridRequestParser consume only the first event and ignore others 😨)
> 
> I'm not sure if it's the best way, but changing signature (without BC break) to return an array could be cleaner
> 
> ```
>     /**
>      * Parses an HTTP Request and converts it into a RemoteEvent.
>      *
>      * @return array<RemoteEvent> Returns empty if the webhook must be ignored
>      *
>      * @throws RejectWebhookException When the payload is rejected (signature issue, parse issue, ...)
>      */
>     public function parse(Request $request, #[\SensitiveParameter] string $secret): array;
> ```

Yes i think only parsing the first message is not an option as you will miss the rest and in our cases allot is bulked together so it would not be worth implementing that i think.

Indeed updating `parse` to be able to return an `array` would be a solution but that will for sure bring complications and will be breaking right. @fabpot maybe you have some insights on what to do next?

TLDR; The problem is that we get multiple message in one webhook request and the current interface allows for only parsing one message in a webhook request.

File: src/Symfony/Component/Mailer/Bridge/Mailchimp/RemoteEvent/MailchimpPayloadConverter.php (hunk 3 above)
Comment (stof) [reply src/Symfony/Component/Mailer/Bridge/Mailchimp/RemoteEvent/MailchimpPayloadConverter.php]: I would say yes. A soft bounce is not deffered

File: src/Symfony/Component/Mailer/Bridge/Mailchimp/RemoteEvent/MailchimpPayloadConverter.php (hunk 2 above)
Comment (JohJohan) [reply src/Symfony/Component/Mailer/Bridge/Mailchimp/RemoteEvent/MailchimpPayloadConverter.php]: Done added the same default case

File: src/Symfony/Component/Mailer/Bridge/Mailchimp/RemoteEvent/MailchimpPayloadConverter.php (hunk 3 above)
Comment (JohJohan) [reply src/Symfony/Component/Mailer/Bridge/Mailchimp/RemoteEvent/MailchimpPayloadConverter.php]: I added `soft_bounce` to `MailerDeliveryEvent::BOUNCE`

File: src/Symfony/Component/Mailer/Bridge/Mailchimp/RemoteEvent/MailchimpPayloadConverter.php (hunk 4 above)
Comment (JohJohan) [reply src/Symfony/Component/Mailer/Bridge/Mailchimp/RemoteEvent/MailchimpPayloadConverter.php]: Now i go over the events and add them all to the reason log with `type` and `diag`

**Review by JohJohan — COMMENTED** [review]
Waiting for https://github.com/symfony/symfony/pull/58248 to be merged then i can update my code for implementation of multiple messages.

**Comment by fabpot** [comment]
@JohJohan #58248 has been merged now.

**Comment by kbond** [comment]
@JohJohan check out https://github.com/symfony/symfony/pull/58403 to see how I implemented for Mailtrap. I suggest adding a `batch.json`/`batch.php` test that contains multiple events.

**Comment by JohJohan** [comment]
@kbond thanks for your implementation of https://github.com/symfony/symfony/pull/58403! i pushed my code i think it should be good now.

**Review by kbond — CHANGES_REQUESTED** [review]
Looks good @JohJohan! Could you add an entry in the bridge's `CHANGELOG.md`?

**Comment by JohJohan** [comment]
> Looks good @JohJohan! Could you add an entry in the bridge's `CHANGELOG.md`?

Thanks for the feedback i updated `src/Symfony/Component/Mailer/Bridge/Mailchimp/CHANGELOG.md` and added:
```md
7.2
---

* Add support for webhook
```

**Review by kbond — APPROVED** [review]

**Review by OskarStark — APPROVED** [review]

File: src/Symfony/Component/Mailer/Bridge/Mailchimp/Webhook/MailchimpRequestParser.php [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+            new MethodRequestMatcher('POST'),
+            new IsJsonRequestMatcher(),
+        ]);
+    }
+
+    protected function doParse(Request $request, #[\SensitiveParameter] string $secret): RemoteEvent|array|null
+    {
+        $content = $request->toArray();
+        if (
+            !empty($content['mandrill_events'])
```
Comment (fabpot) [inline src/Symfony/Component/Mailer/Bridge/Mailchimp/Webhook/MailchimpRequestParser.php]: We're trying to avoid using `empty` in Symfony core. can you remove all their usages in this PR?

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Mailer/Bridge/Mailchimp/Webhook/MailchimpRequestParser.php (hunk 5 above)
Comment (JohJohan) [reply src/Symfony/Component/Mailer/Bridge/Mailchimp/Webhook/MailchimpRequestParser.php]: Good point updated the 3 cases i used `empty`

**Comment by JohJohan** [comment]
Its ready for review, i dont know why https://ci.appveyor.com/project/fabpot/symfony/builds/50744376 fails:

```console
Testing C:\projects\symfony\src\Symfony\Component\Lock
SSSSSSSSSSSSSSSS..............SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS.  63 / 393 ( 16%)
.S.........SSSSSSSSSSSSSSSSSSSSSSSSSSF..........SSSSSSSSSSS.... 126 / 393 ( 32%)
..............SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
Time: 00:50.021, Memory: 8.00 MB
There was 1 failure:
1) Symfony\Component\Lock\Tests\Store\RedisArrayStoreTest::testBackwardCompatibility
Failed asserting that false is true.
C:\projects\symfony\src\Symfony\Component\Lock\Tests\Store\AbstractRedisStoreTestCase.php:49
FAILURES!
Tests: 178, Assertions: 201, Failures: 1, Skipped: 124.
KO src\Symfony/Component/Lock
PHPUnit 9.6.21 by Sebastian Bergmann and contributors.
```

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @JohJohan.
